### ISSUE TITLE
"Already Connected"

### ISSUE THREAD

**Issue by vajnorcan** [Esri/resource-proxy#182]
just playing around with the JAVA proxy page, but when trying to reach some destination through this proxy I get an error "Already Connected". Any ideas what is happening and what shall I change? Map service that I'm trying to reach doens't require any token but it appears like it's still trying to get some..Thanks heaps.

<p>Problem accessing /cbd/proxy.jsp. Reason:
<pre>    Already connected</pre></p><h3>Caused by:</h3><pre>java.lang.IllegalStateException: Already connected
	at sun.net.www.protocol.http.HttpURLConnection.setRequestProperty(HttpURLConnection.java:2378)
	at org.apache.jsp.proxy_jsp.passHeadersInfo(org.apache.jsp.proxy_jsp:138)
	at org.apache.jsp.proxy_jsp._jspService(org.apache.jsp.proxy_jsp:855)
	at org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:109)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:389)
	at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:486)
	at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:380)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:547)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:481)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:119)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:520)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:227)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:940)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:409)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:874)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:117)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:250)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:149)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:110)
	at org.eclipse.jetty.server.Server.handle(Server.java:349)
	at org.eclipse.jetty.server.HttpConnection.handleRequest(HttpConnection.java:441)
	at org.eclipse.jetty.server.HttpConnection$RequestHandler.content(HttpConnection.java:921)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:784)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:223)
	at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:46)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:545)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:43)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:533)
	at java.lang.Thread.run(Thread.java:662)

**Comment by esoekianto** [comment]
@vajnorcan can u specify in which scenario this behavior is happening? for example when making which request specifically? 

I personally never had this error before in Java proxy, but from the error message it looks like there is Input or Output stream that don't get close properly.

**Comment by vajnorcan** [comment]
I'm trying to send the POST request to one of our map services (geometry server, project endpoint). It fails on the "passHeadersInfo" while trying to set the key/value to the connection headers... "con.setRequestProperty" ... I haven't set any tokens/logins.

**Comment by jgravois** [comment]
it'd be helpful to see the actual request (ideally captured using something like Fiddler)

**Comment by vajnorcan** [comment]
if I comment in "passHeadersInfo(request, con);" it works ok. 

*** Request headers ***

Key	Value
Request	POST /cbd/proxy.jsp?https://<some address>/Geometry/GeometryServer/project HTTP/1.1
Accept	*/*
Accept-Language	en-au
Referer	http://localhost:8050/cbd/
x-requested-with	XMLHttpRequest
Content-Type	application/x-www-form-urlencoded
Accept-Encoding	gzip, deflate
User-Agent	Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; rv:11.0) like Gecko
Host	localhost:8050
Content-Length	2278
Connection	Keep-Alive
Cache-Control	no-cache
Cookie	_ga=GA1.1.173141; _gat=1



*** REQUEST BODY ***

f=json&outSR=102100&inSR=3308&geometries=%7B%22geometryType%22%3A%22esriGeometryPolygon%22%2C%22geometries%22%3A%5B%7B%22rings%22%3A%5B%5B%5B9687863.1673%2C4423730.0711%5D%2C%5B9688228.4687%2C4425299.424900001%5D%2C%5B9689269.870800003%2C4425299.424900001%5D%2C%5B9688845.4349%2C4423634.159399999%5D%2C%5B9687863.1673%2C4423730.0711%5D%5D%5D%2C%22spatialReference%22%3A%7B%22wkid%22%3A3308%2C%22latestWkid%22%3A3308%7D%7D%2C%7B%22rings%22%3A%5B%5B%5B9689269.870800003%2C4425299.424900001%5D%2C%5B9690538.298199996%2C4425299.4728%5D%2C%5B9690107.204899996%2C4423441.834100001%5D%2C%5B9689126.426700003%2C4423591.3292%5D%2C%5B9688845.4349%2C4423634.159399999%5D%2C%5B9689269.870800003%2C4425299.424900001%5D%5D%5D%2C%22spatialReference%22%3A%7B%22wkid%22%3A3308%2C%22latestWkid%22%3A3308%7D%7D%2C%7B%22rings%22%3A%5B%5B%5B9687863.1673%2C4423730.0711%5D%2C%5B9688845.4349%2C4423634.159399999%5D%2C%5B9689126.426700003%2C4423591.3292%5D%2C%5B9689104.418300002%2C4423438.7447999995%5D%2C%5B9689245.618%2C4423415.877699999%5D%2C%5B9689153.0136%2C4422866.8662%5D%2C%5B9687716.2738%2C4423095.7094%5D%2C%5B9687863.1673%2C4423730.0711%5D%5D%5D%2C%22spatialReference%22%3A%7B%22wkid%22%3A3308%2C%22latestWkid%22%3A3308%7D%7D%2C%7B%22rings%22%3A%5B%5B%5B9690107.204899996%2C4423441.834100001%5D%2C%5B9689695.072899997%2C4422030.984300001%5D%2C%5B9689073.372599996%2C4422203.5887%5D%2C%5B9688933.431000001%2C4422337.9408%5D%2C%5B9689099.677900001%2C4422507.202199999%5D%2C%5B9689153.0136%2C4422866.8662%5D%2C%5B9689245.618%2C4423415.877699999%5D%2C%5B9689104.418300002%2C4423438.7447999995%5D%2C%5B9689126.426700003%2C4423591.3292%5D%2C%5B9690107.204899996%2C4423441.834100001%5D%5D%5D%2C%22spatialReference%22%3A%7B%22wkid%22%3A3308%2C%22latestWkid%22%3A3308%7D%7D%2C%7B%22rings%22%3A%5B%5B%5B9687716.2738%2C4423095.7094%5D%2C%5B9689153.0136%2C4422866.8662%5D%2C%5B9689099.677900001%2C4422507.202199999%5D%2C%5B9688933.431000001%2C4422337.9408%5D%2C%5B9689073.372599996%2C4422203.5887%5D%2C%5B9689695.072899997%2C4422030.984300001%5D%2C%5B9689428.0167%2C4420989.1756%5D%2C%5B9687648.328900002%2C4421460.061799999%5D%2C%5B9687476.657399997%2C4422137.3719999995%5D%2C%5B9687716.2738%2C4423095.7094%5D%5D%5D%2C%22spatialReference%22%3A%7B%22wkid%22%3A3308%2C%22latestWkid%22%3A3308%7D%7D%5D%7D

*** RESPONSE HEADERS ***

Key	Value
Cache-Control	must-revalidate,no-cache,no-store
Content-Length	4154
Content-Type	text/html;charset=ISO-8859-1
Response	HTTP/1.1 500 Already connected
Server	Jetty(7.5.0.RC2)


Dont get confused by the server. it doesn't work on Tomcat either.

Thanks

**Comment by bsvensson** [comment]
There is no server name in your request:

`Request POST /cbd/proxy.jsp?https:///Geometry/GeometryServer/project`

That doesn't seem right...

**Comment by vajnorcan** [comment]
nah.. I've renamed the server name to <my server name> and it got removed!

**Comment by nickp10** [comment]
I ran into this same error using the Java proxy while submitting a POST request. It's because the forwardToServer method is writing out the body from the POST request and then the passHeadersInfo method is writing out headers. This exception is thrown because you cannot write headers to the connection if you've already begun writing out the body. I fixed it by moving the passHeadersInfo method call inside doHTTPRequest method just before the body is written out.

**Comment by bsvensson** [comment]
@nickp10 - sounds like you have figured out something good.  Do you want to submit it as a pull request?  (If not, we can take care of that...)

**Comment by nickp10** [comment]
Yeah I can submit a pull request.

**Comment by cheetah90** [comment]
@nickp10 I cannot reproduce the error so cannot verify it. To reproduce it, is it just sending POST request to Java/proxy.jsp? When I did it, I receive different error 
```
HTTP Status 500 - An exception occurred processing JSP page /proxy.jsp at line 920

type Exception report

message An exception occurred processing JSP page /proxy.jsp at line 920

description The server encountered an internal error that prevented it from fulfilling this request.

exception

org.apache.jasper.JasperException: An exception occurred processing JSP page /proxy.jsp at line 920

917: 
918: 
919:             String key = "[" + serverUrl.getUrl() + "]x[" + request.getRemoteAddr() + "]";
920:             RateMeter rate = ratemap.get(key);
921:             if (rate == null) {
922:                 rate = new RateMeter(serverUrl.getRateLimit(), serverUrl.getRateLimitPeriod());
923:                 RateMeter rateCheck = ratemap.putIfAbsent(key, rate);


Stacktrace:
	org.apache.jasper.servlet.JspServletWrapper.handleJspException(JspServletWrapper.java:574)
	org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:476)
	org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:396)
	org.apache.jasper.servlet.JspServlet.service(JspServlet.java:340)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
root cause

java.lang.ClassCastException: org.apache.jsp.proxy_jsp$RateMeter cannot be cast to org.apache.jsp.proxy_jsp$RateMeter
	org.apache.jsp.proxy_jsp._jspService(proxy_jsp.java:1036)
	org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:438)
	org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:396)
	org.apache.jasper.servlet.JspServlet.service(JspServlet.java:340)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
note The full stack trace of the root cause is available in the Apache Tomcat/8.0.24 logs.

Apache Tomcat/8.0.24
```

**Comment by cheetah90** [comment]
@nickp10 I cannot reproduce the error so cannot verify it. To reproduce it, is it just sending POST request to Java/proxy.jsp? When I did it, I receive different error 
```
HTTP Status 500 - An exception occurred processing JSP page /proxy.jsp at line 920

type Exception report

message An exception occurred processing JSP page /proxy.jsp at line 920

description The server encountered an internal error that prevented it from fulfilling this request.

exception

org.apache.jasper.JasperException: An exception occurred processing JSP page /proxy.jsp at line 920

917: 
918: 
919:             String key = "[" + serverUrl.getUrl() + "]x[" + request.getRemoteAddr() + "]";
920:             RateMeter rate = ratemap.get(key);
921:             if (rate == null) {
922:                 rate = new RateMeter(serverUrl.getRateLimit(), serverUrl.getRateLimitPeriod());
923:                 RateMeter rateCheck = ratemap.putIfAbsent(key, rate);


Stacktrace:
	org.apache.jasper.servlet.JspServletWrapper.handleJspException(JspServletWrapper.java:574)
	org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:476)
	org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:396)
	org.apache.jasper.servlet.JspServlet.service(JspServlet.java:340)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
root cause

java.lang.ClassCastException: org.apache.jsp.proxy_jsp$RateMeter cannot be cast to org.apache.jsp.proxy_jsp$RateMeter
	org.apache.jsp.proxy_jsp._jspService(proxy_jsp.java:1036)
	org.apache.jasper.runtime.HttpJspBase.service(HttpJspBase.java:70)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:438)
	org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:396)
	org.apache.jasper.servlet.JspServlet.service(JspServlet.java:340)
	javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
note The full stack trace of the root cause is available in the Apache Tomcat/8.0.24 logs.

Apache Tomcat/8.0.24
```

**PS. the same POST request works on PHP and DotNet**

**Comment by bsvensson** [comment]
Closing issue as it has been fixed by PR #239
